### DISCUSSION TITLE
인프라 아키텍처 개선하기 Part.1

### DISCUSSION THREAD

**Discussion by seokhongkim** [woowacourse/infra-architecture-4#2] [General]
팀에서 아래 내용을 함께 논의한 후 정리해서 코멘트로 남겨주세요.

- 개선한 인프라 아키텍처 이미지
- 인프라 아키텍처를 변경하면서 고민한 부분
  - 문제 상황
  - 해결 방법
  - 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분

**Comment by Yboyu0u** [comment]
![image](https://user-images.githubusercontent.com/63579113/193721201-a6f8400c-f799-4d1b-8bdb-8bccbe733653.png)

<**문제 상황>**

- 웹 서버빼고 다 문제다. (WAS, DB)

- 현재 WAS, DB 둘다 하나 밖에 없어서 운영상 매우 위헙한 상황이다. 

<**해결 방법>**

**WAS** 
- 로드 벨런싱을 이용하여 트래픽을 분산하고 한 서버가 죽어도 운영에 지장이 없도록 한다. 

**DB** 
- Replication을 도입해 db가 다운돼도 운영에 지장이 없도록 한다.

<**해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분>**

- 서버를 늘려야 하는 비용이 발생한다.
- 돈을 많이 벌어야 한다.
- 관리 포인트가 늘어난다.
- 설정 같은게 추가될 때 마다 일시적으로 서버가 다운될 수 있다.

**Comment by asebn1** [comment]
![20221004_112657](https://user-images.githubusercontent.com/57438644/193721566-2eaa6e2c-5c0a-4a67-83bb-567f54aa6db1.jpg)

### 문제 상황

- WAS 서버 터짐
- DB서버 터짐
- 퇴사(이직)
- 불의의사고

### 해결 방법

- 문서화
- WAS 서버 이중화
- DB 서버 이중화 (레플리카)
- 무중단 배포

### 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분

- Source DB가 죽었을 땐 어떻게 대처할 것인가
- 무중단 배포시 롤링배포나 카나리 배포는 신버전과 구버전 둘 다 운영되는 구간을 어떻게 관리할 것인가
- 팀 문서화 컨밴션 비용
- 문서와 코드 불일치
- 블루그린 배포는 인스턴스를 많이 필요로하기에 비용이 많이 든다

**Comment by injoon2019** [comment]
![image](https://user-images.githubusercontent.com/46641538/193721434-5870e244-448f-4157-a1a5-38af44a8e903.png)

**문제 상황**
현재 구조에서는 모든 컴포넌트가  SPOF이다. 심지어 Nginx도 문제가 생기면 뒤의 컴포넌트들까지 연결이 되지 않는다. 그래서 ALB를 사용하고, ALB는 문제가 발생하지 않는다고 가정한다.

**해결 방법**
- Nginx
 - Nginx를 두 대로 이중화한다. ALB를 사용해서 Nginx로 로드밸런싱을 한다. 
- WAS
 - 두 대로 이중화한다
- DB
 - Master, Slave 구조로 이중화한다. 

**해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분**
- 배포가 현재 구조보다 복잡해진다
- 서버 비용이 늘어난다
- 관리해야 하는 EC2 가 늘어난다

**Comment by brorae** [comment]
# 공식
![IMG_88C4011F9F18-2](https://user-images.githubusercontent.com/70756680/193722498-ab4912b2-2b21-47a9-b4f9-b782a0793ba8.jpeg)


## 문제 상황

- DB
    - 너무 많이 사용해서 부하가 와서 죽는다.
        - 예비 DB를 둔다.
        - DB 이중화를 한다
        - master slave를 나눌 때 CUD를 할 때 락이 걸리므로 조회용 DB를 따로 둔다.
        - 조회할 때는 조회용 DB만 사용하므로 트래픽이 분산된다.
    - disk가 꽉차서 더 이상 사용할 수 없다.
        - 모니터링하고 스케일업한다.
- WAS
    - 트래픽이 너무 몰려서 서버에 과부하가 올 때
        - 로드 밸런싱을 하여 스케일 아웃을 한다.
            - nginx가 터지면 무의미하다.
            - 로드 밸런싱 알고리즘을 고려해야한다.
- reverse proxy
    - 트래픽이 많아져셔 서버가 다운된다.
        - 스케일 아웃을 한다.

- 트래픽이 너무 많아졌다..!

**Comment by tco0427** [comment]
## 모아모아 인프라 아키텍쳐

---

- 개선한 인프라 아키텍처 이미지

<img width="952" alt="스크린샷 2022-10-04 오전 11 36 26" src="https://user-images.githubusercontent.com/57028386/193722481-99352fc8-b3b0-4244-81fb-22e2d860be1c.png">  

- 인프라 아키텍처를 변경하면서 고민한 부분
    
    **현재 아키텍쳐 구조**

<img width="616" alt="스크린샷 2022-10-04 10 58 00" src="https://user-images.githubusercontent.com/57028386/193722626-546761a5-f056-4005-b7bb-2464ee446438.png">


    - 문제 상황
        - 모든 지점이 SPOF 가 발생할 수 있다.
        - 모든 요청을 하나의 WAS에서 처리하기 때문에 높은 가용성을 제공하지 못한다.
        - DB 인스턴스의 디스크 용량에 제한이 있다.
    - 해결 방법
        - WAS 여러 대를 두고 로드밸런싱을 도입한다.
        - 앞 단의 NGINX 가 두 대 라고 하면 한 대가 다운되도 다른 NGINX를 통해서 요청을 받을 수 있다.
        - DB 같은 곳에 일주일에 한 번 백업하고 전체적으로 크게 백업을 해서 별도의 저장 시스템에 저장한다.
        - 다중화된 WAS 로 인해 DB 사이에 발생하는 병목을 레플리케이션을 통해서 개선한다.
            
            레플리케이션 한 마스터 DB 도 이중화한다. (하나의 마스터 DB가 다운되면 CUD 가 제한되기 때문에)
            
            (Replication을 양방향으로 구성할 수도 있다.)
            
            (싱글 레플리카 복제, 멀티 레플리카 복제, 체인 복제, 듀얼 소스 복제, 멀티 소스 복제)
            
            → 우선 하나의 마스터 DB에서 CUD를 하고 나머지 DB들에서 Read를 한다고 생각하자.
            
    - 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분
        - CRUD 를 모든 DB에 동시에 할 수 있다면 이들간에 트랜잭션은 어떻게 동기화해줄까?
            
            → 분산락?
            
        - 다중화된 서버에 대해 어떻게 배포해줄 수 있을까? 하나씩 배포해야하나?
            
            → 하나씩 배포를 적용하고, 로드 밸런싱에서는 배포된 곳으로 요청을 보낸다.
            
            → 하나씩 더 띄워서 한 번에 갈아 끼운다.
            
            (롤링 방식과 blue&green 방식)
            
        - WAS가 여러개면 synchronized 키워드 붙은 메소드는 각각의 WAS에만 적용되는데, 이는 어떻게 해결할까?
            
            → 어딘가에서 락을 구해와야 한다. DB에서 락을 구해온다?
            
        - 각 DB의 스키마는 어떻게 관리할 것인가?
        - 마스터 DB 를 Active-Active 구조로 구성한다.

**Comment by JangBomi** [comment]
![image](https://user-images.githubusercontent.com/55357130/193722730-045ccf85-0461-4e5b-8df4-55d474b4d1b7.png)


### 문제 상황
- 리버스 프록시 nginx 에 장애가 나면 운영 개발 모두 할 수 없다.
- 유일한 운영 WAS에 장애가 나면 서비스 제공이 불가능하다.
- 유일한 DB에 장애가 나면 운영, 개발 모두 할 수 없다.

### 해결 방법
- 리버스 프록시 nginx를 운영 환경와 개발 환경으로 분리한다.
- 운영 WAS를 다중화하고, 로드밸런서로 연결한다.
- 운영과 개발 DB를 분리한다.
- DB Replication을 도입한다.

### 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분
- 비용이 많이든다.
- 철저한 문서화가 필요하다.
- 관리 포인트가 많아진다.
- 운영 로그는 어떡하지?
- 모든 WAS를 통틀어 한 번만 실행되어야 하는 로직(ex: 알람)을 분리한다.
- 운영 환경과 개발 환경의 인프라 구조가 달라져서 운영 환경에 대한 테스트를 할 수 없다.
- 여러 대의 운영 WAS에 배포가 어떻게 되어야할지 생각해야한다.
- 실제 발생하는 트래픽에 비해 너무 많은 고민을 하고있다 T_T

**Comment by hyeonic** [comment]
<img width="983" alt="image" src="https://user-images.githubusercontent.com/59357153/193723368-8c559f6f-c0e6-41d4-ac69-300c6f86f361.png">


## 문제상황 1. Nginx가 죽는다.

### 해결 방법

1. (HardWare) L4 Switch를 사용하여 로드밸런싱을 한다.
    
![image](https://user-images.githubusercontent.com/59357153/193723003-8a379cd2-b5ee-460a-b983-fb05ed32bb33.png)
    
2. API 별로 별도의 Nginx를 구성하여 프론트에서 요청을 각각의 Nginx로 보낸다.

### 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분

1. 비용이 많이 든다.
2. 도입 가능한지 확실하지 않다. 추측이다. 관리점이 늘어난다.

## 문제상황 2. WAS가 죽는다.

### 해결 방법

1. 서버를 다중화하여 Nginx를 통해 로드 밸런싱 한다.

### 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분

1. 세션을 InMemory를 통해 사용하는 경우 세션 불일치 문제가 발생할 수 있다.
2. DB에서 병목이 발생할 수 있다.

## 문제상황 3. DB가 죽는다.

### 해결 방법

1. 클러스터링
2. 리플리케이션
3. 샤딩

### 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분

1. 클러스터의 스위치가 죽으면, 데이터베이스 전체를 이용할 수 없다.
2. 레플리케이션을 사용하면 데이터 정합성 문제가 발생한다.

**Comment by kbsat** [comment]
![image](https://user-images.githubusercontent.com/52696169/193723011-66004614-7d15-49a6-8889-643969e254b0.png)

### 언제 문제가 될 수 있을까?

- WAS가 터지면 문제가 생긴다.
- DB가 터지면 문제가 생긴다.
- Nginx 서버가 터지면 문제가 생긴다.
- 네트워크 끊기면 문제가 생긴다.
- HTTPS 인증서 만료되면 터짐
- AWS 서울 리전 데이터센터 정전나면 망함
    - 서울 리전에 정전나면
- KT, LG, SKT가 정전나면
- 전쟁나서 인터넷 끊기면 망함
    - 인력이 사라짐
    - 너넨 예비군이라 전방가지? 난 후방간다
        - ㅋㅋ 좋겠다~
- 리전 분산되어 있는 것이 이미 네트워크에 작업이 되어있다 - `gugu클래스피셜`
- 트래픽이 엄청 몰렸을때 터짐

### 어떻게 해결할 수 있을까? 10분

- DB가 터진다
    - DB를 분산한다. 백업서버를 둔다.
        - 데이터베이스 로그를 따로 저장해둔다.
- WAS가 터진다
    - WAS를 다중화
- NGINX가 터진다.
    - Nginx 모니터링 추가
- AWS가 터진다
    - AWS 말고 다른 클라우드도 같이 사용
        - 구글 클라우드 플랫폼
        - 자체 서버
        - 일본 리전 등 가용 영역 분산
        - 다중화된 서버의 서브넷도 다중화
            
            - 어느 서브넷이 터져도 다른 서브넷의 서버에서 연결이 되도록 다중화
            - 우리나라 리전 다터지면 일본 리전 ㄱㄱ
- HTTPS 인증서가 만료된다.
    - HTTPS 만료 전에 미리 확인한다. 끝나는시기를 인지하고있는다.
- 트래픽이 급 몰려온다
    - 동기로 처리하던걸 비동기로 처리할 수 있도록 구조를 개선한다.
        - 메시지큐? 이벤트리스닝? 쿠버네티스?

### 어떤 문제가 생길 수 있을까? 또는 무엇을 일 할때 고려해야 할까? 10분

- WAS를 다중화했는데도 다터지면?
    - 터져도 터져도 끝나지 않는 더 많은 와쓰!!!!!!!!!
    - 내 통장이 터질듯…
- **돈이 많이 든다.. 관리 비용의 증가.. ← 문서화 철저하게해야함**
    - 개발자의 학습비용
- **로그관리가 어려워진다.**
    - 로그를 한 곳에 모은다.
    - 그 로그 관리 서버 터지면 우짬
- **DB가 다중화가 되면 정합성 문제 생길 수 있다.**
    - id가 auto increment는 어떻게 되지?
    - 카뱅의 DB 인프라
        
        [[[135] 오픈소스 데이터베이스, 은행 서비스에 첫발을 내밀다.](https://www.slideshare.net/deview/135-80845610)](https://www.slideshare.net/deview/135-80845610)
        
        - 읽기를 하는 디비와 쓰기를 하는 디비를 분리
        - 쓰기를 하는 마스터 디비는 1개만 두고 읽기를 하는 디비를 여러대 두고 마스터 디비를 복제
- 인프라 구조 인수인계

**Comment by sure-why-not** [comment]
# 땡쿠

<img width="932" alt="스크린샷 2022-10-04 오전 11 47 41" src="https://user-images.githubusercontent.com/94965944/193723742-97e3a528-47e9-4c05-81d2-51828609caab.png">


- 개선한 인프라 아키텍처 이미지
- 인프라 아키텍처를 변경하면서 고민한 부분
    - 문제 상황
        - NginX
            - 인증서 만료
            - WAS 연결이 끊겨거나 요청 오버헤드로 인해 큐에 요청이 계속 쌓일 경우
        - WAS
            - 서버가 죽었을 때 (부하)
        - DB
            - 잘못된 쿼리
            - 트래픽 부하
            - 메모리 부족
        - 개발자
    - 해결 방법
    - 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분
        - 모든 계층이 확장되어야 한다.
        - **NginX**
            - 인스턴스 하나 더 띄워두고 도메인 등록
            - 도커로 관리해서 한번에 환경을 구축하도록 (인증서 등)
        - **WAS**
            - 1차적으론 스케일업을 하고, 그 이후에 스케일 아웃
            - 스케일 아웃을 하면 세션이나 캐시같은 상태를 관리하는 인스턴스를 추가적으로 관리 필요
            - 로그, 스케줄링
            - 로드밸런서 알고리즘
        - DB
            - replication
            - 백업 DB, 로그 저장

**Comment by hamcheeseburger** [comment]
# F12

![KakaoTalk_Photo_2022-10-04-11-45-45](https://user-images.githubusercontent.com/61769743/193723537-5b050429-0df1-4f8e-906a-0a589e412e82.jpeg)

## 문제상황
모든 요청을 1대의 WAS와 DB가 대처한다는 것에 있어서 SPOF 문제가 발생

## 해결방법
마스터 1개에 slave 여러 개 (리플리케이션)
- insert, update, delete(DML)에 대한 것은 master에
- select(READ)에 대한 것은 slave에
- master의 내용을 주기적으로 slave에 sync해준다.
- master가 죽으면 slave가 대체한다.

## 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분

구현은 어떻게 할 수 있을까?

- WAS에서 DML용 READ용 Datasource를 분리하여 관리한다.
- Sync과정에서 문제가 생기면 어떡할까?
    - slave를 내리고 select문도 master가 담당한다.

slave가 1개 더 생기면 어떻게 대처할까?

- WAS의 설정파일이 변경되어야 한다.
- DB와 WAS 사이에 로드 밸런서 혹은 프록시를 위치시킨다.
    - DML은 master DB 서버가 처리하고, READ는 slave DB 서버들이 처리한다.

**Comment by DWL21** [comment]
# 꼭꼭
![IMG_3625](https://user-images.githubusercontent.com/76774809/193722316-619a84bf-b9bd-4067-93af-f92f7faa7112.jpg)

## 문제 상황
- WebServer, WAS, RDBMS중 1개만 작동하지 않아도 서비스를 이용할 수 없다.
- WAS 개수가 늘어나도 Connection Pool 개수가 모자라다.
- 인력 이슈

## 해결 방법
- DBMS ec2는 Master, Slave 구조로 구성한다.
- Master 는 CUD, Slave는 Read를 담당한다.
- Backup DB는 Slave -> 백업DB 담당
- was를 여러개 구성한다.

## 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분
- was가 여러개이면 DB Master, Slave 구조에서 히카리 커넥션 풀 개수를 몇개로 유지해야 하는가?
-  WS가 터지면 대응해야 하나?

**Comment by Seongwon97** [comment]
![image](https://user-images.githubusercontent.com/57744251/193723247-47377c36-5caa-4cda-a99e-4116b5002f28.png)

### 문제 상황
1. Nginx, Spring, MySQL 어디든 터지면 전체 서비스가 터진다..
2. 확장성을 고려하지 않았다.

### 해결 방법
1. Spring을 여러개 띄우고 앞단에 Nginx를 로드밸런싱 용도로 사용
2. MySQL도 여러개의 MySQL을 Master-slave 구조로 둔다.

### 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분
1. 로드밸런싱을 하는 Nginx는 어떻게 처리해야 할까?

**Comment by shindong96** [comment]
## 체크메이트
![아케텍쳐](https://user-images.githubusercontent.com/70707629/193723221-de3ef90e-f825-43d5-b581-ac734024990a.jpeg)

<문제 상황>
- Nginx 다운
- WAS 서버 다운
- DB 서버 다운
- 현재 체크메이트 서비스의 WAS는 스케쥴러를 이용하는 Master, 이용하지 않는 Slave 이중화 구조로 되어 있는데 Master가 다운이 되면 스케쥴러가 동작하지 않음. 

<해결 방법>
- Nginx를 Master, Slave 구조를 이용
- WAS는 이중화로 로드밸런싱한다.
- 각자의 Master, Slave에 배포할 때, Master와 Slave를 교차시켜서 배포한다.
- DB도 Master, Slave로 이중화해서 각자의 DB서버와 Storage를 들고 있고 Master에서 Slave로 replication을 진행한다.

<해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분>
- 관리 비용의 증가
- 배포 방식이 어려워짐
- 꼼꼼한 문서화가 필요
- 로그 분산 문제

**Comment by sojukang** [comment]
<img width="994" alt="인프라" src="https://user-images.githubusercontent.com/83529599/193723695-099a0d07-eedc-4936-8747-89a7366b4c13.png">

## 문제 상황
- Web Server, WAS, DB 모든 계층이 SPOF 이다. 
## 해결 방법
### DB
- Master DB, Slave DB로 다중화
- Master DB에서는 CUD의 Modification 수행
- Agent DB에서는 Read 작업 수행 
- 둘 중 하나에서 장애가 발생하면 나머지 DB에서 모든 작업을 수행 

### WAS
- 두 개의 WAS로 다중화 
- 하나의 WAS는 이전 버전
- 배포한 최신 WAS에 문제가 발생하면 이전 버전의 WAS로 처리

### Web Server 
-  두 개의 Web Server로 다중화

## 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분
- 각 계층에서 두 대를 모두 유지하는 관리, 운영 비용 
- 요청이 몰리는 경우 Load Balancing도 이루어져야 할텐데, 구 버전의 WAS와 신 버전의 WAS를 여러 대 띄워야할까?

**Comment by cjlee38** [comment]
![image](https://user-images.githubusercontent.com/52564093/193723749-8304a9ef-0bb5-4a79-ba97-b1802db69de3.png)

## 문제 상황
- 배포 상황
- EC2 인스턴스가 죽었을 때 

## 해결 방법
- WAS scale out
- DB CQRS
  - 사용자 경험 고려 동기화 처리
- 카나리아 배포

## 고려사항
- Read DB / Write DB 의 각각에 장애가 났을 때 처리방법 ?
- replica는 필요 없을까

**Comment by meatsby** [comment]
![우아한테크코스-7](https://user-images.githubusercontent.com/83967672/193723244-8a0dd335-c4f2-4004-b5ab-c8861533d9d2.jpg)

### 문제 상황
- NginX 장애
- WAS 장애
- DB 장애
### 해결 방법
- NginX 다중화
- WAS 다중화 + Service 별 WAS 분리
- DB Master Slave 분리
### 해결 방법 적용으로 인해 발생하는 이슈와 고려해야 할 부분
- 관리 포인트 증가
- 인력 부족
- DB 동기화

**Comment by soominsohn** [comment]
## 터놓고
**BE** 
열음 애쉬 바니 수달 앤지


**기존 터놓고 인프라 아키텍처**

<img width="532" alt="image" src="https://user-images.githubusercontent.com/83059234/193723820-e598fd1e-65fa-44fa-b54b-8b3bde0e512d.png">

1. NGINX 꺼지면?
    1. nginx가 죽은걸 어떻게 파악할 수 있을까?
    2. 헬스체크 스크립트를 작성한다.
    3. NGINX 두대를 놓는다.
        1. 그러면 요청 분산 어떻게 할건지?
    4. 포트포워딩으로 관리한다.
    5. 밖에서 엔진엔스 헬스체크 해준다.
        1. 포트가 떠있는지 체크해줌.
        2. 죽으면 다른 포트로 띄우겠다.
    6. 다른 리전에다가 새로 띄운다.
    7. nginx dashboard가 있습니다.

1. WAS가 꺼지면?
    1. WAS를 여러대 두어서 로드밸런싱 도입.
    2. nginx가 처리해준다~
    3. WAS끼리 동기화는 어떻게 해야할까?
        1. DB에 낙관락을 건다.

1. DB가 꺼지면?
    1. 백업 DB를 둔다.
    2. Replication DB를 둔다.
        - 단방향 Replication: 읽기전용
        - 양방향 Replication: 데이터 동기화 가능. 읽기/쓰기 둘 다 가능
        - 읽기전용 replication db를 두고 master db가 죽으면?
            - 백업전용 replication db를 두고 백업을 한다.
            - replica 서버 master로 승격시키고, 새로 slave를 만든다.
    3. DB가 죽은걸 어떻게 알 수 있을까?
        1. ping을 보낸다.
    4. MySQL 헬스체크
    5. 샤딩 파티셔닝 등등…

1. EC2가 꺼지면?
    1. 새로 띄운다.
    2. 도커 이미지를 사용한다.
        1. 이미지 어떻게 뜨냐?
        2. 이미지 떠서 도커허브(깃헙 같은거)에 올린다.
        3. 허브에서 이미지 땡겨와서 올린다.

1. AWS에 모니터링 하는 것 없나?
    1. 각자 헬스체크하는 기능이 없나?
    2. 모니터링하는 기술이 있을 것 같다.

1. FE도 여러대 두자.
    1. FE 앞에 nginx 두고 FE WS 여러대 두자.
    2. 그렇다면 FE는 어떤 무중단 배포를 사용해야 할까?
    3. FE - 카나리 배포, BE - 블루그린 배포를 도입했을 때
        1. FE와 BE가 동기화 되지 않은 상황이 발생할 수 있다.
        2. 브라우저 캐시문제도 있다. 
            1. 배포를 할 때 브라우저 캐시를 무효화하는 로직을 넣는다.
    4. 사이에 들어오는 요청을 에러처리 하자.

1. 솔루션
    1. 장애를 파악하는 툴이 있다. (대시보드 같은 것) (nginx로 가능할 듯)
        1. 예) WAS 164 꺼졌어요! 재부팅 해본다.
        2. 그런데 안되면?
        3. 개발자 불러서 장애 복구한다.
    2. 근데 nginx가 꺼지면?
        1. 자동화가 가능할까?
        2. 모니터링 툴을 도입한다.
            1. ec2 헬스체크
            2. log 분석
    3. 모든 것을 자동화 할 수 없다면 사람의 개입을 최소화하는 방향으로 개선한다.
    

**개선해 본 점**

<img width="567" alt="image" src="https://user-images.githubusercontent.com/83059234/193723849-acdb1800-05e2-472e-a626-ecd5b63fb874.png">

**Reply by her0807** [reply]
디비 복제시 분산락 , 세선 정보도 하나의 와스에 저장되니까 세션 관리하는 방법도 ..
DB 를 여러대 쓴다고 했을 때, Connection 관리 방법에 대해서도 ..

**Reply by her0807** [reply]
디비 이중화에 대해서는 was transaction 동기화에 대한 고민도 해보기

**Comment by tonic523** [comment]
### 스모디

![image](https://user-images.githubusercontent.com/59171113/193723059-1cf50258-64fe-4240-8401-e7989f3c2eef.png)

## 문제가 터지는 상황
- Client에서 요청하는 Nginx가 터지는 경우
- WAS가 터지는 경우
- DB가 터지는 경우
- Spring으로 구현된 Image 서버가 터지는 경우

하나의 서버에 장애가 발생하면 서비스가 중단되는 문제가 발생한다.

## 해결 방법
- nginx를 여러 개 둔다. nginx 서버에 문제가 발생하면 예비 도메인 주소를 가지는 nginx로 갈아끼운다.
- nginx에서 로드벨런싱을 구현하여 여러 개의 WAS로 트래픽을 분산한다.
- DB는 레플리케이션을 통해 백업 DB로 동기화를 시켜준다.
- 이미지 서버에 저장되는 이미지 파일들에 대한 백업 서버를 만들어 동기화를 해준다.

## 해결 방법 중 명확하지 않은 부분
- nginx 서버에 문제가 발생했을 때 매번 domain 호스팅 서버에 가서 백업 nginx의 ip로 수정해주어야 한다.
- 레플리케이션을 통해 백업 DB를 만들었을 때 WAS에서 백업 DB로 url을 바꿀 수 있는지
- 이미지 파일을 동기화하는 방법
